### ISSUE TITLE
Expose tracemalloc C API to track/untrack memory blocks

### ISSUE THREAD

**Issue by vstinner** [python/cpython#74240]
BPO | [30054](https://bugs.python.org/issue30054)
--- | :---
Nosy | @vstinner
PRs | <li>python/cpython#1236</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-06-20.15:47:31.122>
created_at = <Date 2017-04-12.13:42:25.341>
labels = ['type-feature', '3.7']
title = 'Expose tracemalloc C API to track/untrack memory blocks'
updated_at = <Date 2017-06-20.15:47:31.121>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-06-20.15:47:31.121>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-06-20.15:47:31.122>
closer = 'vstinner'
components = []
creation = <Date 2017-04-12.13:42:25.341>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 30054
keywords = []
message_count = 7.0
messages = ['291554', '291556', '291557', '291558', '292029', '296463', '296464']
nosy_count = 2.0
nosy_names = ['vstinner', 'jtaylor']
pr_nums = ['1236']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue30054'
versions = ['Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
The issue bpo-26530 added a private C API to manually track/untrack memory blocks in tracemalloc. I was just validated by Julian Taylor who confirms that the API works as expected:
http://bugs.python.org/issue26530#msg291551

So I propose to make the 3 newly added functions public and document them:

* \_PyTraceMalloc_Track()
* \_PyTraceMalloc_Untrack()
* \_PyTraceMalloc_GetTraceback()

**Comment by 503c56e2-f019-4587-ba44-17ad2b762efa** [comment]
I am not sure if \_PyTraceMalloc_GetTraceback really needs to be a public function.
Exposing the tracing information should probably just go over python interfaces.

**Comment by vstinner** [comment]
What's the status in numpy? Is the current numpy release able to use CPython 3.6 tracemalloc private functions?

**Comment by 503c56e2-f019-4587-ba44-17ad2b762efa** [comment]
With this changeset it would:
https://github.com/numpy/numpy/pull/8885

**Comment by vstinner** [comment]
@jtaylor: Ok, I created https://github.com/python/cpython/pull/1236

**Comment by vstinner** [comment]
New changeset 5ea4c0677389ead2eee759958694cff2c65834a7 by Victor Stinner in branch 'master':
bpo-30054: Expose tracemalloc C API (bpo-1236)
https://github.com/python/cpython/commit/5ea4c0677389ead2eee759958694cff2c65834a7

**Comment by vstinner** [comment]
I merged my PR: the future Python 3.7 will provide a public C API to track memory allocations.
